README.md

```diff
@@ -1,2 +1,105 @@
 # unified-user-data-format
+
 Specification for the unified user data format across the various "Tube" apps (NewPipe, FreeTube, Invidious, etc..)
+
+## Intro
+
+* This specification is a very early draft, and prone to regular changes
+* Unless otherwise specified, all fields are assumed to be `Strings`
+
+
+## Top level structure
+
+#### `"version"`
+
+The [semantic version](https://semver.org/) of the specification.
+
+Conditions for bumping major/minor/patch version: TBD
+Potential suffixes: `-draft`, TBD
+
+#### `"subscriptions"`
+
+An array of Subscription objects. See below for details.
+
+#### `"playlists"`
+
+An array of Playlist objects. See below for details.
+
+### Overview
+
+```javascript
+{
+	"version": "X.Y.Z[-draft/-rc]",
+	"subscriptions": Subscriptions[],
+	"playlists": Playlist[],
+}
+```
+
+## Subscription object
+
+#### > `"name"`
+
+The name of the subscribed channel.
+
+#### > `"url"`
+
+The URL used to access the user subscribed channel.
+
+#### > `"thumbnail"`
+
+The URL to the user/channel's thumbnail.
+
+### Example
+
+```javascript
+{
+  "name": "Linus Tech Tips",
+  "url": "https://www.youtube.com/channel/UCXuqSBlHAE6Xw-yeJA0Tunw",
+  "thumbnail": "https://yt3.ggpht.com/ytc/abcdef..."
+}
+```
+
+## Playlist object
+
+#### > `"name"`
+
+The title of the playlist.
+
+#### > `"visibility"`
+
+The visibility of the playlist.
+
+Accepted values: "public", "unlisted", "private"
+
+#### > `"description"`
+
+The description of the playlist.
+Format: plain text (? TBD)
+
+CAN be ommited; In this case, the parser MUST assume an empty `String`.
+
+#### > `"videos"`
+
+An array of `String`.
```

Comment:
Could this instead be an object with and `id` and `url` for parsing simplicity?

Reply:
@SamantazFox  https://en.wikipedia.org/wiki/Uniform_Resource_Name#Namespaces
> In order to ensure the global uniqueness of URN namespaces, their identifiers (NIDs) are required to be registered with the [IANA](https://en.wikipedia.org/wiki/Internet_Assigned_Numbers_Authority).

Reply:
We don't really need to pack information in only one string in a specific way. We can just define a json object of this form to replace URLs:
```json
{
  "type": "video",
  "service": "youtube",
  "instance": null,
  "id": "djV11Xbc914",
  "url": "https://www.youtube.com/watch?v=djV11Xbc914",
}
```

Also, if we decide not to use URLs, we should agree on what exactly an ID is for all services. For example, the first letters of YouTube playlist IDs are used to specify the playlist type, so technically those should not appear in `"id"`, but rather many different playlist `"type"`s should be created. Another example can be observed with PeerTube's accounts vs channels: they are (slightly) different entities with different URLs, so either we distinguish them into two different `"type"`s, or we include part of the URL in the ID. Examples: [YouTube normal playlist (PL)](https://www.youtube.com/playlist?list=RDCLAK5uy_nZiG9ehz_MQoWQxY5yElsLHCcG0tv9PRg), [YouTube genre mix (RDCLAK)](https://www.youtube.com/playlist?list=RDCLAK5uy_nZiG9ehz_MQoWQxY5yElsLHCcG0tv9PRg), [PeerTube account (/a/)](https://framatube.org/a/programmingfish@tube.piweb.be), [PeerTube channel (/c/)](https://framatube.org/c/programmingfish_channel@tube.piweb.be).

I think creating new `"type"`s each time a service is added is complicated, breaks compatibility and provides almost useless information. Therefore, only three types should exist: channel, playlist and video. The ID field should then contain standardized IDs based on the service (so e.g. PeerTube channel/account IDs would be, respectively, `c/...` and `a/...`).

... or we can just use only URLs directly and avoid all of these complications. In the end URLs are already "Uniform Resource Locator"s, so why bother creating our own way to uniquely locate resources? In this case we should also agree on which URL formats are ok, e.g. we might decide that these things are forbidden: useless parameters (e.g. t=5), short urls (e.g. youtu.be), http instead of https, YouTube urls from other frontends (e.g. https://invidious.kavin.rocks/watch?v=djV11Xbc914 instead of https://www.youtube.com/watch?v=djV11Xbc914)... Maintaining such a list of requirements is not so difficult, and even if one requirement were not respected by a frontend, other frontends would be able to open the URLs anyway in most cases, since usually frontends do allow opening strange/random URLs from the user anyway.

Reply:
I personally prefer the json object approach rather than a url, as libraries for interacting with frontends/services typically do not have you pass a url, rather you may call high order functions with parameters to specify what data to return. if a url was used a (likely more complicated) parser would be required.

When it comes to the id, it may differ depending on what library is being used, so it is probably best to see what each of your libraries use.

Reply:
Maybe we could include the `url` along with the `id`?  That seems like a more reasonable approach to me. I do agree that it can be tedious at times to parse just the id.

We can't just include just the `id`, as that contains no information about the service it is for.

Reply:
In the example json object posted:
> ```
> {
>  "type": "video",
>  "service": "youtube",
>  "instance": null,
>  "id": "djV11Xbc914",
>  "url": "https://www.youtube.com/watch?v=djV11Xbc914",
>}
>```

The service, instance and type are specified.